docker-compose-development.yml

```diff
@@ -0,0 +1,51 @@
+# This file is configured for deployment of the CantusDB project in local development.
+# When building the project locally, replace the contents of docker-compose.yml with the
+# contents of this file.
+
+version: '3'
+
+services:
+  django:
+    build:
+      context: ./django
+    volumes:
+      - ./:/code/
+      - static_volume:/resources/static
+      - media_volume:/resources/media
+      - api_cache_volume:/resources/api_cache
+    env_file: ./config/envs/dev_env
+    ports:
+      - 3000:3000
+    restart: always
+    depends_on:
+      - postgres
+
+  nginx:
+    build:
+      context: ./nginx
+    ports:
+      - 80:80
+      # - 443:443
```

Comment:
Perhaps this could be an environment variable.

Reply:
Are you suggesting that we include this line (or don't include it) based on an environment variable, or are you saying we can include a list of port mappings in an environment variable?

I've taken a look at CU's `docker-compose.yaml`, and done some cursory googling, and can't see how we can include/exclude a variable based on a boolean environment variable. Do you have something specific in mind?

Reply:
A good argument for figuring out how to do this via environment variable is that it would allow us to have just a single `docker-compose.yml` file, yes?

Reply:
I'm saying would it work to have a variable in `dev_env` that is `PORT` and then have this be:

```yaml
ports:
     - ${PORT}:${PORT}
```

so that this can be set appropriately in the dev_env to whatever environment you're in (`443` on a server, `80` for development)

Reply:
> A good argument for figuring out how to do this via environment variable is that it would allow us to have just a single `docker-compose.yml` file, yes?

Yes

Reply:
> I'm saying would it work to have a variable in `dev_env` that is `PORT` and then have this be:
> 
> ```yaml
> ports:
>      - ${PORT}:${PORT}
> ```
> 
> so that this can be set appropriately in the dev_env to whatever environment you're in (`443` on a server, `80` for development)

Ah, I understand! And this would be specified in a new line in our `dev_env` file, correct?

Reply:
I've tried a few different ways to specify this environment variable in `dev_env`, but I keep getting an error when I try to `docker-compose build`:

```
$ docker-compose build
WARN[0000] The "PORT" variable is not set. Defaulting to a blank string. 
WARN[0000] The "PORT" variable is not set. Defaulting to a blank string. 
1 error(s) decoding:

* error decoding 'Ports': No port specified: :<empty>
```

(concluding that the variables in `dev_env` only show up once the container is built. So we would need to set an environment variable _outside_ the container before trying to build...?)

Reply:
Env variables can be passed from a file to Docker Compose: https://docs.docker.com/compose/environment-variables/set-environment-variables/

Reply:
yes - our docker-compose.yml already includes this line in the django container, and I added it to our nginx container too while testing this. Let me try again...

Reply:
looking at your message again and the documentation, the env_file is a parameter to the `docker-compose build` command, and not to be included in `docker-compose.yml` itself. Got it.

Reply:
I wonder whether this would actually make the configuration process simpler.

(Perhaps this is standard procedure and I've just not become accustomed to it having spent a bunch of time going through CantusDB's current deployment workflow as my only interaction with docker-compose...)

Reply:
> yes - our docker-compose.yml already includes this line in the django container, and I added it to our nginx container too while testing this. Let me try again...

Did this not work?

Reply:
Correct - including `env_file: ./config/envs/dev_env` in the `nginx` block did not resolve the error that `- ${PORT}:${PORT}` was causing.

As far as I can tell, specifying an `env_file` means that the variables defined in that file will be available inside the container when it is running (allowing django, for example, to [get the value of `CANTUSDB_DEBUG` in `settings.py`](https://github.com/DDMAL/CantusDB/blob/13e91741d06999a80879c745a8e75319460d7fdd/django/cantusdb_project/cantusdb/settings.py#L39)). But it doesn't make the variables available to docker when it's evaluating `docker-compose.yml` itself.

Doing some digging, it appears that we can specify all the necessary environment variables in the command every time we call `docker-compose` (i.e., `docker-compose up -d` becomes `PORT=80 docker-compose up -d`). But if we do this, I'm worried about it cascading into a bunch of changes in several other files, especially those relating to our cron setup.

(reference: https://serversforhackers.com/c/div-variables-in-docker-compose . This suggests that maybe the error I'm hitting could be resolved by putting `dev_env` directly in the root of the repository, next to `docker-compose.yml`?)

Reply:
> Part of this page suggests that maybe the error I'm hitting could be resolved by putting `dev_env` directly in the root of the repository

This is what CU does and it works, but it's annoying that you can't specify another path... I guess it was worth a try? But I agree that at this point it's more trouble than it's worth (and adding things to the command, I agree, seems like it would be potentially troublesome)